### ISSUE TITLE
FlipBook advanceTime() never completes

### ISSUE THREAD

**Issue by marcojakob** [bp74/StageXL#178]
The `FlipBook.advanceTime(...)` method never returns `false` and thus never indicates that it is completed. See https://github.com/bp74/StageXL/blob/master/lib/src/display_ex/flip_book.dart#L102

This makes the `Juggler` always keep the `FlipBook` in its animatable list.

**Comment by bp74** [comment]
You are right, the question is what is the right condition to return false? If you call the stop or pause method it doesn't necessary mean that you want to remove the flipbook from the juggler. You have to add the flipbook to the juggler yourself, so maybe the right thing would be to remove it on your own to? I don't know.

**Comment by bp74** [comment]
Maybe an optional parameter to the stop method which indicates if the flipbook should be removed from the juggler too?

**Comment by marcojakob** [comment]
I have a more complex set up: 

1. Create an `AnimationGroup`.
2. Add some animations including the `FlipBook` to the `AnimationGroup`.
3. Add the `AnimationGroup` to the juggler.

After a single `FlipBook` loop the `AnimationGroup` should be marked as completed and removed from the juggler. But since `FlipBook` never completes, this never happens.

I don't know what the best way is to design this. Maybe add an optional parameter to `play()` that destroys the flipbook when the animation ends (but only if loop=false).

The phaser.io library has a similar approach in its Sprite.Animatable.play() method. There is a killOnComplete parameter: http://docs.phaser.io/AnimationManager.js.html

But it's a bit different in phaser.io because they implement flipbook animations directly inside the sprites.

**Comment by marcojakob** [comment]
I just realized that I'm probably using `FlipBook` in a wrong way. I was removing the Sprite on animation start and replacing it with the flipbook. At the end of the animation I was adding the sprite again.

Well, I should probably use the flipbook all the time and set it to the start frame when not animating.

**Comment by bp74** [comment]
If the first frame is the same as the Bitmap you are using in your Sprite, this would probably be a good idea. The render performance of a Flipbook is the same as a Bitmap, so you won't loose any performance. If you go to frame 0 and stop the Flipbook the advanceTime call from the Juggler won't take much time too. Keep me posted about your progress and i will think about the Flipbook and some automatic Juggler removal too.

**Comment by marcojakob** [comment]
I've kept the complicated setup for now because it wasn't as easy as I thought with using the flipbook only. I realized that I need many different flipbook animations: for moving the character in each direction and for turning.

---

### My thoughts about the `FlipBook` class

* `FlipBook` implements `advanceTime()` from `Animatable`. It says that `advanceTime()` returns `false` if it is completed. `FlipBook.advanceTime()` fires a completeEvent when the last frame is reached (and loop is false) but it never returns false and thus does not indicate that it completed. So, this is kind of confusing that it completes but doesn't actually complete. I think that we should return false at the moment the complete event is fired.

* I would **not** complete the `FlipBook` when it is stopped. Actually, I think that `stop()` should be renamed to `pause()`. This would express more accurately what it does: It pauses the animation to later be continued from the same frame. With stop some might expect it to automatically reset back to the first frame and maybe even fire a complete event. 

* I think `FlipBook.advanceTime()` might need some refactoring; it is very difficult to reason about the behavior of the `while (_play)` loop. As far as I can tell the two checks for `_currentFrame != nextFrame` can never be `true` because a few lines above the `_currentFrame` is set to the `nextFrame`:   
```dart
      while (_play) {

        var frameDuration = _frameDurations[_currentFrame];
        var lastFrame = _currentFrame;
        var nextFrame = _loop ? (lastFrame + 1) % totalFrames : min(lastFrame + 1, totalFrames - 1);

        if (_frameTime < frameDuration) break;

        _currentFrame = nextFrame;
        _frameTime -= frameDuration;

        // dispatch progress event on every new frame
        if (lastFrame != nextFrame) {
          this.dispatchEvent(_progressEvent);
          if (_currentFrame != nextFrame) return true;
        }

        // dispatch complete event only on last frame
        if (lastFrame != nextFrame && nextFrame == totalFrames - 1 && _loop == false) {
          this.dispatchEvent(_completeEvent);
          if (_currentFrame != nextFrame) return true;
        }
      }
```

### My thoughts about the `Animatable` class hierarchy

* In `FlipBook`, the `onProgress` and `onComplete` events are actually `Streams`. I think this makes sense but is somewhat inconsistent with how other `Animatable` classes handle it. Classes like `Tween` or `Transition` use a callback `Function` instead of `Streams`. I'd prefer to have `Streams` everywhere as this is more *Darty* than callbacks. I know that changing this would mean a breaking change so maybe we could wait for *async-await* to be officially supported and then switch everything over to *async-await* (will it land in Dart 1.9? I'm not sure).

* I'm not sure about this, but wouldn't it make sense for all `Animatables` to have an `onStart`, `onUpdate` and `onComplete` stream? If yes, we could move those up to the `Animatable` interface.

**Comment by bp74** [comment]
Sorry Marco for not responding to this topic for some time. I was quite busy doing other things but starting with next week my head is free for other stuff again and i will start working on your suggestions.

**Comment by marcojakob** [comment]
No problem, I've been on vacation myself last week :-)
